spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java

```diff
@@ -0,0 +1,156 @@
+    // For more information on WriteStream.Type, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1/WriteStream.Type.html
+    WriteStream stream = WriteStream.newBuilder().setType(WriteStream.Type.PENDING).build();
+
+    CreateWriteStreamRequest createWriteStreamRequest =
+        CreateWriteStreamRequest.newBuilder()
+            .setParent(parentTable.toString())
+            .setWriteStream(stream)
+            .build();
+    WriteStream writeStream = client.createWriteStream(createWriteStreamRequest);
+
+    // Use the JSON stream writer to send records in JSON format.
+    // For more information about JsonStreamWriter, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1beta2/JsonStreamWriter.html
+    streamWriter =
+        JsonStreamWriter.newBuilder(writeStream.getName(), writeStream.getTableSchema()).build();
+  }
+
+  /**
+   * @param data JSONArray to be appended
+   * @param offset offset at which data has to be added
+   */
+  public void append(JSONArray data, long offset)
+      throws DescriptorValidationException, IOException, ExecutionException {
+    synchronized (this.lock) {
+      // If earlier appends have failed, we need to reset before continuing.
+      if (this.error != null) {
+        throw this.error;
+      }
+    }
+    // Append asynchronously for increased throughput.
+    ApiFuture<AppendRowsResponse> future = streamWriter.append(data, offset);
+    ApiFutures.addCallback(future, new AppendCompleteCallback(), MoreExecutors.directExecutor());
+    // Increase the count of in-flight requests.
+    inflightRequestCount.register();
+  }
+
+  /**
+   * Call this method before committing the stream
+   *
+   * @param client BigQueryWriteClient
+   */
+  public void finalizeWriteStream(BigQueryWriteClient client) {
+    // Wait for all in-flight requests to complete.
+    inflightRequestCount.arriveAndAwaitAdvance();
+
+    // Verify that no error occurred in the stream.
+    synchronized (this.lock) {
+      if (this.error != null) {
+        throw this.error;
+      }
+    }
+
+    // Finalize the stream.
+    FinalizeWriteStreamResponse finalizeResponse =
+        client.finalizeWriteStream(streamWriter.getStreamName());
+    logger.info("\nRows written: " + finalizeResponse.getRowCount());
+  }
+
+  public String getStreamName() {
+    return streamWriter.getStreamName();
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Close the connection to the server.
+    streamWriter.close();
+  }
+
+  class AppendCompleteCallback implements ApiFutureCallback<AppendRowsResponse> {
+    private final Logger logger = LoggerFactory.getLogger(AppendCompleteCallback.class);
+
+    public void onSuccess(AppendRowsResponse response) {
+      logger.info(
+          String.format("\nAppend %d success", response.getAppendResult().getOffset().getValue()));
```

Comment:
String.format() has a cost, but as it happens, slf4j `logger` has overloaded method variants that take a string pattern and parameters. This way, the string substitution will only be done if `INFO` level is needed.

Reply:
done